### ISSUE TITLE
"no relay/exit support" when running default executable

### ISSUE THREAD

**Issue by tdietert** [GaloisInc/haskell-tor#24]
This is the output I get when running the default exe generated by a successful build.

[2016-03-26 12:16] Credentials created.
[2016-03-26 12:16]   Signing key fingerprint: 8134b7fba2aa271884c7ccf9ac4314b056f68784
[2016-03-26 12:16]   Onion key fingerprint:   6cb4771e317699b5f06497a18488314131a78787
[2016-03-26 12:16]   TLS key fingerprint:     b111b5ce1c241663f8d6c8c22a7c16cc96c55cfa
[2016-03-26 12:16] Fetching credentials for default directory moria1 [128.31.0.39:9131]
[2016-03-26 12:16] Fetch failed: Fetch timed out.
[2016-03-26 12:16] Fetching credentials for default directory tor26 [86.59.21.38:80]
[2016-03-26 12:16] Fetching credentials for default directory dizum [194.109.206.212:80]
[2016-03-26 12:16] Fetching credentials for default directory Tonga [82.94.251.203:80]
[2016-03-26 12:16] Fetch failed: Parser error: Failed reading: HTTP Error: Not found ["\r\n"]
[2016-03-26 12:16] Fetching credentials for default directory gabelmoo [131.188.40.189:80]
[2016-03-26 12:16] Fetching credentials for default directory dannenberg [193.23.244.244:80]
[2016-03-26 12:16] Weird: fingerprint mismatch. Ignoring dir.
[2016-03-26 12:16] Fetching credentials for default directory Faravahar [154.35.175.225:80]
[2016-03-26 12:16] Fetching credentials for default directory longclaw [199.254.238.52:80]
[2016-03-26 12:16] 5 of 8 default directories loaded.
[2016-03-26 12:16] String consensus document update.
[2016-03-26 12:16] Using directory dizum for consensus.
[2016-03-26 12:16] Waiting for Tor connections on port 9002
[2016-03-26 12:16] Added new directory entry for dannenberg
[2016-03-26 12:16] Added new directory entry for maatuska
[2016-03-26 12:16] Added new directory entry for urras
[2016-03-26 12:17] Failed to add new directory for moria1
[2016-03-26 12:17] Found enough valid signatures: dannenberg, tor26, longclaw, maatuska, urras, dizum, gabelmoo, Faravahar
[2016-03-26 12:17] New router processing complete. 2950 of 6177 routers available.
[2016-03-26 12:17] Will reload census at 13:53
[2016-03-26 12:17] Trying to connect to 84.75.230.35
[2016-03-26 12:17] Just built connection with them.
[2016-03-26 12:17] Created new link to 84.75.230.35 ("OdyX")
[2016-03-26 12:17] Built initial link to "84.75.230.35" with circuit ID 4172906913
[2016-03-26 12:17] Created circuit 4172906913
[2016-03-26 12:17] Extending circuit to "78.22.238.222"
[2016-03-26 12:17] Extending circuit to "176.9.4.206"
[2016-03-26 12:17] Extending circuit to "14.136.236.129"
[2016-03-26 12:17] Extending circuit to "86.219.120.200"
[2016-03-26 12:17] Extending circuit to exit node "94.228.86.11"
[2016-03-26 12:17] I believe I have the following addrs: [IP4 "149.18.82.104"]
[2016-03-26 12:17] Trying to connect to 62.143.98.61
[2016-03-26 12:17] Just built connection with them.
[2016-03-26 12:17] Created new link to 62.143.98.61 ("Unnamed")
[2016-03-26 12:17] Built initial link to "62.143.98.61" with circuit ID 3743545541
[2016-03-26 12:17] Created circuit 3743545541
[2016-03-26 12:17] Extending circuit to "212.47.254.99"
[2016-03-26 12:17] Extending circuit to "166.70.207.2"
[2016-03-26 12:17] Extending circuit to "212.227.20.116"
[2016-03-26 12:17] Extending circuit to "176.9.148.176"
[2016-03-26 12:17] Extending circuit to exit node "198.50.200.139"
[2016-03-26 12:17] Failed to create connection to myself. No relay/exit support.

Do I have to configure a port on my laptop to act as an exit node? Thanks in advance!

**Comment by acw** [comment]
Well, regarding the first set, I've had some timeout problems with directory servers before, so that's not too unusual.

With regard to the "failed to create a connection to myself" error, that's fine, too, as long as you don't want to be a relay or exit node. :) You'll get that error when either one of your firewalls blocks incoming connections to your address from the Internet, your NAT'd without the appropriate forwarding rules, or (more annoyingly) when something in the circuit (typically the exit node) flakes out on you. Are any of those conditions likely?

**Comment by tdietert** [comment]
Ahh, so let me confirm that this isn't necessarily unexpected behavior? 

I'm pretty sure I'm not behind a firewall (home router) and I'm unsure how to check if the net addr was translated with the proper forwarding rules... can you tell me where I might read more about this? I'm also unsure how to diagnose whether the exit node failed or not; does the execution output I listed above give any insight into this? If I run the program 5 times in a row, I get pretty much the same output every time.

**Comment by acw** [comment]
Nope, it can happen. Try running it a couple times; it should get through once.

Also, upon a second look, it appears that it did correctly generate a circular route:

`[2016-03-26 12:17] I believe I have the following addrs: [IP4 "149.18.82.104"]`

So really, it's just a matter of a really bad message in an exception handler.
